designs/RFC - Network Setup Teardown Optimization.md

```diff
@@ -0,0 +1,396 @@
+    %10 = comp.schedule_read %arg3 from %1 on %0 wait for %7 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    comp.wait %8, %9, %10 : !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>
+  }
+  comp.destroy_kernel %k1 : !comp.kernel<ocl, 11, 8x32xf32>
+  comp.destroy_kernel %k2 : !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.dealloc %0 %4 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>) -> ()
+  comp.dealloc %0 %5 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>) -> ()
+  comp.dealloc %0 %1 : (!comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>) -> ()
+  comp.destroy_execenv %0 : !comp.execenv<ocl:0,(11)>
+  return
+}
+```
+
+### Final Lowering
+
+With a minor tweak, we could run the above code as a coroutine &mdash; we could pass in an extra parameter to represent
+a bidirectional request/response pipe, and pass it to `comp.loop`.  Our sense is that it's more obvious to split the
+current `main()` function into multiple functions invoked individually to set up the network, run it, and tear it down.
+
+To do this, we augment the existing `!comp.execenv` to serve as a resource container, and add accessor operations for
+the various resources.  For our example code, this transformation looks something like this:
+
+```mlir
+func @plaidml_init(%arg0: !comp.device) -> !comp.execenv<ocl:0,11)> {
+  %0 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
+  %1 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<8x32xf32, 11>
+  comp.execenv_set_alloc %0, %1 {midx = 0} : !comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>
+  %4 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<16x32xf32, 11>
+  comp.execenv_set_alloc %0, %4 {midx = 1} : !comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>
+  %5 = comp.alloc %0 : (!comp.execenv<ocl:0,(11)>) -> memref<8x16xf32, 11>
+  comp.execenv_set_alloc %0, %5 {midx = 2} : !comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>
+  %k1 = comp.create_kernel %0 {kernel = @main_kernel::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 8x32xf32>
+  comp.execenv_set_kernel %0, %k1 {kidx = 0} : !comp.execenv<ocl:0,(11)>, !comp.kernel<ocl, 11, 8x32xf32>
+  %k2 = comp.create_kernel %0 {kernel = @main_kernel_0::@main_kernel} : (!comp.execenv<ocl:0,(11)>) -> !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.execenv_set_kernel %0, %k2 {kidx = 1} : !comp.execenv<ocl:0,(11)>, !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  return %0
+}
+
+func @plaidml_exec(%arg0: !comp.execenv<ocl:0,11>, %arg1: memref<8x16xf32>, %arg2: memref<16x32xf32>, %arg3: memref<8x32xf32>) {
+  %c2 = constant 2 : index
+  %c32 = constant 32 : index
+  %c1 = constant 1 : index
+  %c8 = constant 8 : index
+  %c4 = constant 4 : index
+  %1 = comp.execenv_get_alloc %arg0 {midx = 0}: (!comp.execenv<ocl:0,(11)>) -> memref<8x32xf32, 11>
```

Comment:
I thought we were going to use values instead of using `midx` and `kidx`?

Reply:
Misunderstand on my end; I thought it was just the way the code was using those to smuggle values.  Again, I think I can eliminate it; it'll be in the next update.